File: ruby/lib/basecamp/oauth/token.rb
# frozen_string_literal: true

module Basecamp
  module Oauth
    # OAuth 2 access token response.
    #
    # @attr access_token [String] The access token string
    # @attr token_type [String] Token type (usually "Bearer")
    # @attr refresh_token [String, nil] The refresh token string
    # @attr expires_in [Integer, nil] Lifetime of the access token in seconds
    # @attr expires_at [Time, nil] Calculated expiration time
    # @attr scope [String, nil] OAuth scope granted
    Token = Data.define(
      :access_token,
      :token_type,
      :refresh_token,
      :expires_in,
      :expires_at,
      :scope
    ) do
      def initialize(
        access_token:,
        token_type: "Bearer",
        refresh_token: nil,
        expires_in: nil,
        expires_at: nil,
        scope: nil
      )
        # Calculate expires_at from expires_in if not provided
        calculated_expires_at = expires_at || (expires_in ? Time.now + expires_in : nil)
        super(
          access_token: access_token,
          token_type: token_type,
          refresh_token: refresh_token,
          expires_in: expires_in,
          expires_at: calculated_expires_at,
          scope: scope
        )
      end

      # Checks if the token is expired or about to expire.
      #
      # @param buffer_seconds [Integer] Buffer time before actual expiration (default: 60)
      # @return [Boolean] true if expired or will expire within buffer time
      def expired?(buffer_seconds = 60)
        return false unless expires_at

        Time.now + buffer_seconds >= expires_at
      end
    end
  end
end

File: python/src/basecamp/oauth/token.py
from __future__ import annotations

import time
from dataclasses import dataclass, field


@dataclass(frozen=True)
class OAuthToken:
    """OAuth 2 access token response."""

    access_token: str
    # RFC 6750 authentication scheme name, not a credential.
    token_type: str = "Bearer"  # noqa: S105
    refresh_token: str | None = None
    expires_in: int | None = None
    expires_at: float | None = field(default=None)
    scope: str | None = None
    #: RFC 8707 resource indicator the token is bound to (BC5:
    #: ``urn:bc:account:<id>``). Echo it as the ``resource`` parameter when
    #: refreshing — BC5 multi-account refresh tokens reject a refresh
    #: without it (SPEC §16). Appended last: earlier fields keep their
    #: positional slots.
    resource: str | None = None

    def __post_init__(self) -> None:
        # Calculate expires_at from expires_in when not explicitly provided.
        if self.expires_at is None and self.expires_in is not None:
            object.__setattr__(self, "expires_at", time.time() + self.expires_in)

    def is_expired(self, buffer_seconds: int = 60) -> bool:
        """Check if the token is expired or will expire within *buffer_seconds*."""
        if self.expires_at is None:
            return False
        return time.time() + buffer_seconds >= self.expires_at
